### PR TITLE
Create executeOnDestContextByCaller

### PR DESCRIPTION
**Pull request by camilbancioiu**
Create EEI method `executeOnDestContextByCaller()`, and simplify the methods `executeOnDestContext()`, `executeOnSameContext()`, `delegateExecution()` and `executeReadOnly()`.

### CODE REVIEW

File: arwen/elrondapi/elrondei.go [hunk 1]
```diff
@@ -1732,10 +1647,105 @@ func getOriginalTxHash(context unsafe.Pointer, dataOffset int32) {
...
+	}
+
+	currOffset := dataOffset
+	data, err := arwen.GuardedMakeByteSlice2D(numArguments)
+	if err != nil {
+		return "", nil, 0, err
+	}
+
+	for i := int32(0); i < numArguments; i++ {
+		currArgLenData := argumentsLengthData[i*4 : i*4+4]
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:1730]: i know this is copied from old code - but can we make it too look better ?

File: arwen/host/arwen.go [hunk 2]
```diff
@@ -372,6 +372,15 @@ func (host *vmHost) hasRetriableExecutionError(vmOutput *vmcommon.VMOutput) bool
 	return vmOutput.ReturnMessage == "allocation error"
 }
 
+// AreInSameShard returns true if the provided addresses are part of the same shard
+func (host *vmHost) AreInSameShard(leftAddress []byte, rightAddress []byte) bool {
```
Comment (sasurobert) [inline arwen/host/arwen.go:376]: is this called always by sender SC and receiver SC ? hopefully not sender address and SC.

File: mock/vmHostMock.go [hunk 3]
```diff
@@ -78,6 +78,10 @@ func (host *VmHostMock) ExecuteOnDestContext(_ *vmcommon.ContractCallInput) (*vm
 	return nil, nil, nil
 }
```
Comment (sasurobert) [inline mock/vmHostMock.go:80]: comments ?

File: mock/vmHostStub.go [hunk 4]
```diff
@@ -136,6 +137,13 @@ func (vhs *VmHostStub) ExecuteOnDestContext(input *vmcommon.ContractCallInput) (
 	return nil, nil, nil
 }
 
+func (vhs *VmHostStub) AreInSameShard(left []byte, right []byte) bool {
```
Comment (sasurobert) [inline mock/vmHostStub.go:140]: comments ?

**Review by sasurobert — CHANGES_REQUESTED** [review]
need a contract with test

File: arwen/elrondapi/elrondei.go (hunk 1 above)
Comment (camilbancioiu) [reply arwen/elrondapi/elrondei.go]: Yes. Refactored `getArgumentsFromMemory()` and added new function `runtime.MemLoadMultiple()`.

File: arwen/host/arwen.go (hunk 2 above)
Comment (camilbancioiu) [reply arwen/host/arwen.go:376]: There was a situation where `AreInSameShard()` was indeed called incorrectly, between the original TX caller and the receiver SC. Fixed.

File: mock/vmHostMock.go (hunk 3 above)
Comment (camilbancioiu) [reply mock/vmHostMock.go:80]: Added.

File: mock/vmHostStub.go (hunk 4 above)
Comment (camilbancioiu) [reply mock/vmHostStub.go:140]: Added.

File: mock/vmHostMock.go (hunk 3 above)
Comment (camilbancioiu) [reply mock/vmHostMock.go:80]: Not yet - there is an upcoming PR that refactors the mocks and stubs: https://github.com/ElrondNetwork/arwen-wasm-vm/pull/205

Will make sure that comments will be added there.

File: mock/vmHostStub.go (hunk 4 above)
Comment (camilbancioiu) [reply mock/vmHostStub.go:140]: Not yet - see above.

**Comment by camilbancioiu** [comment]
Added test contracts for the success case.

**Review by sasurobert — APPROVED** [review]

File: arwen/contexts/runtime.go [hunk 5]
```diff
@@ -711,6 +711,26 @@ func (context *runtimeContext) MemLoad(offset int32, length int32) ([]byte, erro
 	return result, nil
 }
 
+func (context *runtimeContext) MemLoadMultiple(offset int32, lengths []int32) ([][]byte, error) {
```
Comment (BeniaminDrasovean) [inline arwen/contexts/runtime.go:714]: Missing comment.

File: arwen/host/testInitializer.go [hunk 6]
```diff
@@ -37,12 +37,19 @@ func GetSCCode(fileName string) []byte {
...
-// GetTestSCCode retrieves the bytecode of a WASM testing module
+// GetTestSCCode retrieves the bytecode of a WASM testing contract
 func GetTestSCCode(scName string, prefixToTestSCs string) []byte {
 	pathToSC := prefixToTestSCs + "test/contracts/" + scName + "/output/" + scName + ".wasm"
 	return GetSCCode(pathToSC)
 }
 
+// GetTestSCCodeByName retrieves the bytecode of a WASM testing contract, given
+// a specific name of the WASM module
+func GetTestSCCodeModule(scName string, moduleName string, prefixToTestSCs string) []byte {
```
Comment (BeniaminDrasovean) [inline arwen/host/testInitializer.go:48]: Update comment.

**Review by BeniaminDrasovean — DISMISSED** [review]

File: arwen/contexts/runtime.go (hunk 5 above)
Comment (camilbancioiu) [reply arwen/contexts/runtime.go:715]: Added.

File: arwen/host/testInitializer.go (hunk 6 above)
Comment (camilbancioiu) [reply arwen/host/testInitializer.go:48]: Fixed.

**Review by BeniaminDrasovean — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**